doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst

```diff
@@ -525,25 +525,33 @@ This section describes some special purpose tactics to work with
 
 .. tacn:: discriminate {? @induction_arg }
 
-   Proves any goal for which a hypothesis in the form :n:`@term__1 = @term__2`
-   states an impossible structural equality for an inductive type.
-   If :n:`@induction_arg` is not given, it checks all the hypotheses
+   Proves goals for which a hypothesis in the form :n:`@term__1 = @term__2`
+   has inconsistent constructors between the two sides of
+   the equality.  For example, :g:`(S (S O)) = (S O)` is impossible.
+   It relies on the fact that constructors of inductive types are injective,
+   i.e. if `C1` and `C2` are constructors of an inductive type then
+   :n:`C1 @term__1 <> C2 @term__2`.
+
+   The tactic traverses the normal forms of :n:`@term__1` and :n:`@term__2`,
+   looking for subterms placed in the same positions and whose
+   head symbols are different constructors. If such subterms are present, the
+   equality is impossible and the current goal is completed.
+   Otherwise the tactic fails.  Note that opaque constants are not expanded by
+   δ reductions while computing the normal form.
```

Comment:
The last sentence does not make sense to me.

```suggestion
   Otherwise the tactic fails.
```

Talking about "normal form" is strange, because `discriminate` does not compute it (I certainly hope so), it just computes the head normal form. Moreover, `discriminate` does not care about opacity (unless it refers to `Qed`, but this is not the place to talk about the difference between `Defined` and `Qed`); it will delta-reduce as much as needed to get the head constructors.

Reply:
Normal form is already mentioned in the first sentence: "The tactic traverses the **_normal forms_** of term1 and term2 ...."

It looks like `discriminate` calls `Reductionops.nf_betaiota`
Call chain:
 `Equality.discr_tac` ->
`Equality.discrEverywhere` ->
`Equality.discr` ->
`Equality.onEquality` ->
`Tacred.reduce_to_quantified_ind` ->
`Tacred.reduce_to_ind_gen` ->
`Reductionops.nf_betaiota`

So "not expanded by δ reductions" seems correct.  I don't know how helpful the sentence is to the user, but I think I'd keep it (not fully grasping its significance).  Feel free to suggest different wording.  WDYT?

Also, I tweaked the first paragraph.